atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/QosCassandraWriteTestSuite.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.atlasdb.ete;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.IntStream;
+
+import org.assertj.core.util.Lists;
+import org.junit.After;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.rules.RuleChain;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.Uninterruptibles;
+import com.palantir.atlasdb.containers.CassandraEnvironment;
+import com.palantir.atlasdb.http.errors.AtlasDbRemoteException;
+import com.palantir.atlasdb.todo.ImmutableTodo;
+import com.palantir.atlasdb.todo.Todo;
+import com.palantir.atlasdb.todo.TodoResource;
+
+public class QosCassandraWriteTestSuite extends EteSetup {
+    private static final List<String> CLIENTS = ImmutableList.of("ete1");
+
+    @ClassRule
+    public static final RuleChain COMPOSITION_SETUP = EteSetup.setupComposition(
+            QosCassandraWriteTestSuite.class,
+            "docker-compose.qos.cassandra.yml",
+            CLIENTS,
+            CassandraEnvironment.get());
+
+    @Test
+    public void shouldFailIfWritingTooManyBytes() {
+        TodoResource todoClient = EteSetup.createClientToSingleNode(TodoResource.class);
+
+        ensureOneWriteHasOccurred(todoClient);
+
+        assertThatThrownBy(() -> todoClient.addTodo(getTodoOfSize(100_000)))
+                .isInstanceOf(RuntimeException.class)
+                .as("Cannot write 100_000 bytes the second time as write limit of "
+                        + "1000 was consumed and the burst isnt enough either.");
+    }
+
+    private void ensureOneWriteHasOccurred(TodoResource todoClient) {
+        try {
+            todoClient.addTodo(getTodoOfSize(100_000));
+            // okay as the first huge write is not rate limited.
+        } catch (Exception e) {
+            // okay as some other test might have written before
```

Comment:
we should just reset the txn manager between tests to avoid this (with the current impl, you could also just change the rate supplier, since the limits get reset on any change)

Reply:
Interesting.